### PR TITLE
Convert to file-scope namespaces

### PR DESCRIPTION
**Pull request by MikeAlhayek**
Yes this will cause a conflict for every PR we have. However, the current PRs seems to have no activities for a while.

If any PR need to fix it's merge conflicts, simply convert the conflicted files to file-scope namespaces or use this command

> dotnet format --severity=info --diagnostics=IDE0160 IDE0161

Note to reviewers, since it would be crazy to review 3.5k files, I run the following command

> dotnet format --severity=info --diagnostics=IDE0160 IDE0161

Then made the changes to the following files:
- `.editorconfig` changed the `csharp_style_namespace_declarations` to warning.
- updated the PR CI workflow so it verify if the format is correct based on the preferences in `.editorconfig`

### CODE REVIEW

**Comment by Piedone** [comment]
As with other such PRs, it only makes sense if you configure an analyzer to guard against it, and make sure it actually fails the build in CI.

**Comment by MikeAlhayek** [comment]
We already have a suggestion for this in the `.editorconfig` https://github.com/OrchardCMS/OrchardCore/blob/5c45566274b96dae775af260c3b3214d418ee2e1/.editorconfig#L35 maybe we can use `warn` instead of `suggestion`

**Comment by MikeAlhayek** [comment]
@Piedone I changed the editor settings to warn against using block-scope namespaces.

**Comment by Piedone** [comment]
Please check if the CI build indeed fails.

**Comment by sebastienros** [comment]
> Yes this will cause a conflict for every PR we have. However, the current PRs seems to have no activities for a while.

There is a way to prevent that, apply the same change in the files that are changed by every PR. Would take a script to do it, but it doesn't seem very hard, even more with the gh cli.

- list all PRs
- foreach PR
  - apply the same tool (and changes)
  - merge main (there shouldn't be any conflicts since the untouched files would be identical)
  - push to original PR (even on fork)

**Comment by MikeAlhayek** [comment]
> Please check if the CI build indeed fails.

Seems to be passing.

**Comment by MikeAlhayek** [comment]
> There is a way to prevent that, apply the same change in the files that are changed by every PR. Would take a script to do it, but it doesn't seem very hard, even more with the gh cli.

This approach sounds promising. But, I don't have much time to adventure with this approach. If you have the time, adding this script will be time saver for the PR.

**Comment by Piedone** [comment]
> > Please check if the CI build indeed fails.
> 
> Seems to be passing.

I mean, it should fail if there's a non-compliant file. Changing the .editorconfig doesn't guarantee you'll see the same behavior in CI.

**Comment by hishamco** [comment]
I remembered Seb against this change in the past, because it will break every PR, it would be nice to add this incrementally and new PRs

**Comment by MikeAlhayek** [comment]
@Piedone I added CI to ensure important formatting is apply on the PR CI.

File: .editorconfig
Comment (Piedone) [inline .editorconfig:35]: Locally, everything should be at most a warning, so you can ignore them while you're prototyping. Then, due to `TreatWarningsAsErrors`, the CI build will still fail.

File: .gitattributes
Comment (Piedone) [inline .gitattributes:4]: Why are these necessary?

File: .github/workflows/pr_ci.yml
Comment (Piedone) [inline .github/workflows/pr_ci.yml:37]: This shouldn't be necessary under Linux: https://git-scm.com/book/en/v2/Customizing-Git-Git-Configuration

File: src/OrchardCore.Build/OrchardCore.Commons.props
Comment (Piedone) [inline src/OrchardCore.Build/OrchardCore.Commons.props:16]: This will use the analyzer shipped with the .NET SDK, i.e. use the version that one has locally. This can cause inconsistent results, since analyzers change (a lot, you can see dozens of new violations in a previously pristine codebase with a new patch version even). A better way is to reference the analyzer package directly ([example](https://github.com/Lombiq/.NET-Analyzers/blob/5506480624315c9db6eccefe0a4097a81c357494/Lombiq.Analyzers/AnalyzerPackages.props#L10)).

However, `RunAnalyzersDuringBuild` ([see](https://learn.microsoft.com/en-us/visualstudio/code-quality/disable-code-analysis?view=vs-2022#net-framework-projects), and contrary the docs referring to .NET FW it also affects .NET) needs to be set to `false` by default, and overridden in the command in CI builds. This is so when you build locally, all the analyzers won't run for the whole solution (you'll still see the violations in the IDE for the currently open files, as well as during build for changed files).

File: .github/workflows/pr_ci.yml
Comment (Piedone) [inline .github/workflows/pr_ci.yml:39]: This shouldn't be necessary, due to the build causing warnings to fail.

File: .github/workflows/pr_ci.yml
Comment (Piedone) [inline .github/workflows/pr_ci.yml:35]: This shouldn't really do anything, since it's after the build that already restores.

**Review by Piedone — CHANGES_REQUESTED** [review]
This would be good, but indeed with some help for PR authors, at least with a comment with guidelines under each (I think just "1. `dotnet format` 2. commit/push and don't be afraid that all files changed 3. merge" should be enough).

File: .editorconfig
Comment (MikeAlhayek) [reply .editorconfig:35]: I am not sure that dotnet format uses TreatWarningsAsErrors which is why this was needed. Ill try it with warning

File: .gitattributes
Comment (MikeAlhayek) [reply .gitattributes:4]: because running `dotnet format` command failed without .csproj atleast

File: .github/workflows/pr_ci.yml
Comment (MikeAlhayek) [reply .github/workflows/pr_ci.yml:37]: without this the command `dotnet format` will create differences and fail.

File: .editorconfig
Comment (Piedone) [reply .editorconfig:35]: That's for `dotnet build`. `dotnet format` isn't needed.

File: .github/workflows/pr_ci.yml
Comment (MikeAlhayek) [reply .github/workflows/pr_ci.yml:39]: without this, there is no guarantee that the formatting is correct,

File: .github/workflows/pr_ci.yml
Comment (MikeAlhayek) [reply .github/workflows/pr_ci.yml:35]: yes. I think this was done for a test and left.

File: .github/workflows/pr_ci.yml
Comment (Piedone) [reply .github/workflows/pr_ci.yml]: Analyzers, including code style analyzers, integrate into the build too, causing build warnings/errors, so can fail the build without `dotnet format` (you only need that if you want to apply formatting, or verify formatting without a build).

**Comment by MikeAlhayek** [comment]
@Piedone look at my last commit. I expected the build to fail since the `ITask` does not use file-scope namespaces.

image

**Comment by Piedone** [comment]
See https://github.com/OrchardCMS/OrchardCore/pull/16539#discussion_r1709911933

File: src/OrchardCore.Build/OrchardCore.Commons.props
Comment (MikeAlhayek) [reply src/OrchardCore.Build/OrchardCore.Commons.props]: @Piedone I don't think I understand. I already removed the added `EnforceCodeStyleInBuild` then we build the project using this command `dotnet build -c Release -warnaserror /p:TreatWarningsAsErrors=true /p:RunAnalyzers=true /p:NuGetAudit=false`

so I would expect the the build to fail but instead it passes.

File: .github/workflows/pr_ci.yml
Comment (MikeAlhayek) [reply .github/workflows/pr_ci.yml]: The idea behind `dotnet format` is to run it and see if there is changes "just like we do with assets validation".

Comment (Piedone) [reply .github/workflows/pr_ci.yml]: Unlike asset validation, we don't need it though :). We need the build to take care of this, because then it's one less moving part, and easier to reproduce the violations in any environment, and it'll be supported by any IDE.

This is not some kind of hack BTW, analyzers are meant to run as part of the build, this is a standard, recommended approach. It's also necessary for all analyzers not used by `dotnet format`.

**Comment by lahma** [comment]
Maybe also consider [ignore revisions configuration for Git](https://github.blog/changelog/2022-03-24-ignore-commits-in-the-blame-view-beta/) after this has been merged: https://github.com/sebastienros/jint/commit/a08f4528ef2c3a370f12962183ac275d6865ce36

**Comment by MikeAlhayek** [comment]
> Maybe also consider [ignore revisions configuration for Git](https://github.blog/changelog/2022-03-24-ignore-commits-in-the-blame-view-beta/) after this has been merged: https://github.com/sebastienros/jint/commit/a08f4528ef2c3a370f12962183ac275d6865ce36

@lahma that sounds like a good idea. But I think we have to do it after we merge this PR to get the hash or the merge-squash commit so we can add it to the `.git-blame-ignore-revs` file. Correct?

**Comment by lahma** [comment]
> > Maybe also consider [ignore revisions configuration for Git](https://github.blog/changelog/2022-03-24-ignore-commits-in-the-blame-view-beta/) after this has been merged: [sebastienros/jint@a08f452](https://github.com/sebastienros/jint/commit/a08f4528ef2c3a370f12962183ac275d6865ce36)
> 
> @lahma that sounds like a good idea. But I think we have to do it after we merge this PR to get the hash or the merge-squash commit so we can add it to the `.git-blame-ignore-revs` file. Correct?

Yes, I believe you need to merge this first and then add he file when you know the exact hash in the repo.

**Review by Piedone — CHANGES_REQUESTED** [review]

File: .github/workflows/pr_ci.yml
Comment (Piedone) [reply .github/workflows/pr_ci.yml]: I fixed the build behavior. Having to double-declare IDE* rules is also something we painstakingly learned with https://github.com/Lombiq/.NET-Analyzers.

Please do any clean-up necessary (like I suppose none of the currently commented-out code in the CI workflow is needed) and make sure that `OrchardCore.Commons.props` is the right file to put these into (I'm not sure; it looks like it, but the hierarchy of these props files, and how they reference each other, is complex).

File: .gitattributes
Comment (MikeAlhayek) [reply .gitattributes]: done

File: .github/workflows/pr_ci.yml
Comment (MikeAlhayek) [reply .github/workflows/pr_ci.yml]: Thanks! Everything now looks find and should be where needs to be. the CI is working as expected now.

**Review by Piedone — APPROVED** [review]
Please help PR authors though: https://github.com/OrchardCMS/OrchardCore/pull/16539#pullrequestreview-2228371245

File: .gitattributes
Comment (sebastienros) [reply .gitattributes]: text=auto is supposed to do the conversion automatically. I don't think it's a good idea to force crlf.

**Comment by MikeAlhayek** [comment]
I think all the anthers needs to do is run the following command in their PRs

> dotnet format --severity=info --diagnostics=IDE0160 IDE0161

Then maybe also `gulp rebuild` if they made any modifications to assets.

We would have to merge this one, then try the above steps on some of the PRs to see if that fixes everything or if we need to other steps.

The problem is that most of the current PRs already have merge conflicts. and trying to fix them with the above two commands may not be easy. But either way, they should start with the above commands, and they fix the per-existing merge conflict.

File: .gitattributes
Comment (sebastienros) [reply .gitattributes]: And each extension can be forced for a specific format (text or binary)

Examples: 
https://github.com/dotnet/dotnet/blob/main/src/aspire/.gitattributes
https://github.com/dotnet/dotnet/blob/main/src/aspnetcore/.gitattributes

Comment (Piedone) [reply .gitattributes]: It's a Roslyn bug: https://github.com/dotnet/roslyn/issues/55526

**Comment by Piedone** [comment]
Don't tell me, tell the PR authors :).

**Comment by MikeAlhayek** [comment]
> Don't tell me, tell the PR authors :).

After we merge this, I'll attempt to fix a PR and provide steps for the others to follow.

File: .gitattributes
Comment (sebastienros) [reply .gitattributes]: I read the comments, and they don't seem to imply there is a bug, I don't think so either. Sam correctly mentions the end_of_line should not be forced and that's it:

It also needs to remove this from .editorconfig and there won't be conflicting behaviors.

```
end_of_line = crlf
```

Comment (sebastienros) [reply .gitattributes]: Mike did the changes, the CI is green. I also cloned on Linux after these changes, did a dotnet format and no files were updated.

And it's LF:

```
00000000: 7573 696e 6720 5379 7374 656d 3b0a 7573  using System;.us
00000010: 696e 6720 5379 7374 656d 2e43 6f6c 6c65  ing System.Colle
00000020: 6374 696f 6e73 2e47 656e 6572 6963 3b0a  ctions.Generic;.
00000030: 7573 696e 6720 5379 7374 656d 2e43 6f6c  using System.Col
00000040: 6c65 6374 696f 6e73 2e49 6d6d 7574 6162  lections.Immutab
```

**Review by sebastienros — APPROVED** [review]

**Comment by MikeAlhayek** [comment]
@Piedone if yoy are good with the last changes, please merge this PR. Otherwise, please provide more feedback.

File: .gitattributes
Comment (Piedone) [reply .gitattributes]: We'll see :).

Comment (sebastienros) [reply .gitattributes]: There is a possibility there will be some glitches, but I think only from files that are incorrectly tagged as text vs. binary, or some which are currently cr-lf but should be lf only (web assets for instance).

I can have a look at one of the repositories where you have had the issue if you want and see if I can understand why you are getting problems.

Comment (Piedone) [reply .gitattributes]: It seems to me the recommendation is to enforce the same line ending everywhere and not let the platform handle it: https://github.com/dotnet/roslyn/issues/55526#issuecomment-1165001300 I don't particularly like this, but we're actually doing that in the Lombiq repos (including `end_of_line = crlf` in the .editorconfig) and it works reliably, no issues since then.

But perhaps indeed if we removed all of this then it'd also work.

This was a problem with all of our projects using https://github.com/Lombiq/.NET-Analyzers, i.e. where [these props and config files](https://github.com/Lombiq/.NET-Analyzers/tree/dev/Lombiq.Analyzers) are loaded. Thank you!

**Pull request merged**
